### PR TITLE
Add known limitation about schema changes between executions of prepared statements

### LINKED ISSUE #1511: sql: error on result type changes in prepared statements
1.0 known limitation

PR: https://github.com/cockroachdb/cockroach/pull/16335#issuecomment-306293183

### PR DESCRIPTION
**Pull request by jseldess**
Fixes #1511

### CODE REVIEW

File: known-limitations.md [hunk 1]
```diff
@@ -248,3 +248,33 @@ Given a query like `SELECT * FROM foo WHERE a > 1 OR b > 2`, even if there are a
 ## Privileges for `DELETE` and `UPDATE`
 
 Every [`DELETE`](delete.html) or [`UPDATE`](update.html) statement constructs a `SELECT` statement, even when no `WHERE` clause is involved. As a result, the user executing `DELETE` or `UPDATE` requires both the `DELETE` and `SELECT` or `DELETE` and `UPDATE` [privileges](privileges.html) on the table.
+
+## Schema changes between executions of prepared statements
+
+When the schema of a table targeted by a `PREPARE` statement changes before the prepared statement is executed, CockroachDB should return an error. Instead, CockroachDB incorrectly allows the prepared statement to return results based on the changed table schema, for example:
```
Comment (mjibson) [inline known-limitations.md]: We need to note that this affects both kinds of prepare statements. I'm not sure this is the clearest way, but maybe:

> When the schema of a table targeted by a prepared statement or a `PREPARE` statement changes, CockroachDB should return an error the next time it is executed.

File: known-limitations.md [hunk 2]
```diff
@@ -248,3 +248,33 @@ Given a query like `SELECT * FROM foo WHERE a > 1 OR b > 2`, even if there are a
...
+~~~
++----+-----------+
+| id |   name    |
++----+-----------+
+|  1 | Max Roach |
++----+-----------+
+(1 row)
+~~~
+
+Also, a prepared [`INSERT`](insert.html) or [`UPSERT`](upsert.html) statement executed via the pgwire protocol (not via `PREPARE` and `EXECUTE` keywords) acts inconsistently when the schema of the table being written to is changed before the prepared statement is executed:
```
Comment (mjibson) [inline known-limitations.md]: This also affects DELETE.

File: known-limitations.md [hunk 3]
```diff
@@ -248,3 +248,33 @@ Given a query like `SELECT * FROM foo WHERE a > 1 OR b > 2`, even if there are a
...
+
+- If the number of columns has increased, the prepared statement returns an error but nonetheless writes the data.
+- If the number of columns remains the same but the types have changed, the prepared statement writes the data and does not return an error.
```
Comment (mjibson) [inline known-limitations.md]: This point is true for both pgwire and `PREPARE`.

Comment (jseldess) [inline known-limitations.md]: Again, I've just made this more generic. Look ok?

File: known-limitations.md (hunk 2 above)
Comment (jseldess) [inline known-limitations.md]: Done.

File: known-limitations.md (hunk 1 above)
Comment (jseldess) [inline known-limitations.md]: Since this applies to both, I've just make it more generic. PTAL.

**Pull request merged**